File: python/day3.py
import sys

lines = list(sys.stdin)
ss = [l.split(" ") for l in lines]

rects = []

for s in ss:
    print(s)
    x, y = s[2][:-1].split(',')
    w, h = s[3].split('x')
    x, y, w, h = int(x), int(y), int(w), int(h)

    x2, y2 = x + w, y + h
    rects.append((x, y, x2, y2))

cloth = [[0 for _ in range(1000)] for _ in range(1000)]

for r in rects:
    x, y, x1, y1 = r
    for i in range(x, x1):
        for j in range(y, y1):
            cloth[i][j] += 1

print(sum(c > 1 for r in cloth for c in r))

ids = [s[0] for s in ss]

for idx, r in enumerate(rects):
    x, y, x1, y1 = r
    if all(cloth[1] <= 1 for i in range(x, x1) for j in range(y, y1)):
        print(idx[idx])

File: rust/src/bin/day3.rs
use std::io::prelude::*;
use std::io;
use std::collections::HashMap;

#[derive(Default)]
struct Rect {
    id: i32,
    pos: Vec<i32>,
    dim: Vec<i32>,
}

fn main() {
    let stdin = io::stdin();
    let mut map = HashMap::new();
    let rects: Vec<Rect> = stdin.lock()
        .lines()
        .map(|line| {
            let line = line.unwrap();
            let mut r: Rect = Default::default();
            let line: Vec<&str> = line.split(' ').collect();
            match &line[..] {
                &[id, _, coords, dim] => {
                    r.id = id[1..].parse().unwrap();
                    r.pos = coords[..coords.len() - 1].split(',')
                        .map(|s| s.parse().unwrap())
                        .collect();
                    r.dim = dim.split('x')
                        .map(|s| s.parse().unwrap())
                        .collect();
                }
                _ => {}
            }
            r
        })
        .collect();

    for r in rects {
        let (x, y, w, h) = (r.pos[0], r.pos[1], r.dim[0], r.dim[1]);

        for i in x..x + w {
            for j in y..y + h {
                let space = map.entry((i, j)).or_insert(0);
                *space += 1;
            }
        }
    }

    let overlap = map.values().filter(|v| **v > 1).count();
    println!("{}", overlap);
}
